### ISSUE TITLE
ArrayOutOfBoundsException in SourceTaintingMV

### ISSUE THREAD

**Issue by ahmayun** [gmu-swe/phosphor#197]
# Problem 
## Summary
I'm having trouble trying to get Phosphor running on a toy program. The setup (i.e. instrumenting JRE and the toy program) works fine for me without errors. It's only when I specify a taint source that a runtime error occurs. I'll describe the problem first and then describe the setup process I followed later.

## Toy Program (to be instrumented)

In this program, I want the `taintSource` function to be the taint source and the `taintSink` function to be the sink. 

```
public class TaintExample {
        public static int taintSource(int i) {
                return i;
        }

        public static void taintSink(int i) {
                System.out.println("Sink");
        }

        public static void main(String[] args) {
                int num1 = taintSource(Integer.parseInt(args[0]));
                int num2 = taintSource(Integer.parseInt(args[1]));
                int sum = num1 + num2;

                System.out.println("The sum of " + num1 + " and " + num2 + " is: " + sum);
                taintSink(sum);
        }
}
```
I package this into a jar file and instrument it using the following command:
```
$:~ java -jar Phosphor/target/Phosphor-0.1.0-SNAPSHOT.jar TaintExample.jar inst
```

This creates a jar with instrumented classes in `inst/TaintExample.jar`. I know this works fine because I've observed the bytecode in `inst/TaintExample.jar` and it has all of phosphor's instrumentation.

## Sources and Sinks
I create two files `TaintExample.sources` and `TaintExample.sinks` with the following contents:

TaintExample.sources:
> ```
> TaintExample.taintSource(I)I
> ```
TaintExample.sinks:
> ```
> TaintExample.taintSink(I)V
> ```

## The Error

When I run the following command I get an `ArrayIndexOutOfBoundsException`. I also get the correct output at the end.
```
$:~ jre-inst/bin/java -javaagent:phosphor-jigsaw-javaagent/target/phosphor-jigsaw-javaagent-0.1.0-SNAPSHOT.jar=taintSources=TaintExample.sources,taintSinks=TaintExample.sinks -jar inst/TaintExample.jar 12 12

java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 1
        at java.base/edu.columbia.cs.psl.phosphor.instrumenter.SourceTaintingMV.autoTaintArguments(SourceTaintingMV.java:87)
        at java.base/edu.columbia.cs.psl.phosphor.instrumenter.SourceTaintingMV.visitLdcInsn(SourceTaintingMV.java:66)
        at java.base/edu.columbia.cs.psl.phosphor.org.objectweb.asm.ClassReader.readCode(ClassReader.java:2398)
        at java.base/edu.columbia.cs.psl.phosphor.org.objectweb.asm.ClassReader.readMethod(ClassReader.java:1491)
        at java.base/edu.columbia.cs.psl.phosphor.org.objectweb.asm.ClassReader.accept(ClassReader.java:721)
        at java.base/edu.columbia.cs.psl.phosphor.org.objectweb.asm.ClassReader.accept(ClassReader.java:401)
        at java.base/edu.columbia.cs.psl.phosphor.SourceSinkTransformer.transform(SourceSinkTransformer.java:64)
        at java.base/edu.columbia.cs.psl.phosphor.PhosphorBaseTransformer.signalAndTransform(PhosphorBaseTransformer.java:21)
        at edu.columbia.cs.psl.jigsaw.phosphor.javaagent.PreMain$PhosphorTransformerBridge.transform(PreMain.java:36)
        at edu.columbia.cs.psl.jigsaw.phosphor.javaagent.PreMain$PhosphorTransformerBridge.transform(PreMain.java:26)
        at java.instrument/java.lang.instrument.ClassFileTransformer.transform(ClassFileTransformer.java:244)
        at java.instrument/sun.instrument.TransformerManager.transform(TransformerManager.java:188)
        at java.instrument/sun.instrument.InstrumentationImpl.transform(InstrumentationImpl.java:565)
        at java.instrument/sun.instrument.InstrumentationImpl.transform(InstrumentationImpl.java)
        at java.instrument/sun.instrument.InstrumentationImpl.retransformClasses0(Native Method)
        at java.instrument/sun.instrument.InstrumentationImpl.retransformClasses0(InstrumentationImpl.java)
        at java.instrument/sun.instrument.InstrumentationImpl.retransformClasses(InstrumentationImpl.java:168)
        at java.instrument/sun.instrument.InstrumentationImpl.retransformClasses(InstrumentationImpl.java)
        at edu.columbia.cs.psl.jigsaw.phosphor.javaagent.PreMain$1.retransformClasses(PreMain.java:68)
        at java.base/edu.columbia.cs.psl.phosphor.SourceSinkTransformer.retransform(SourceSinkTransformer.java:114)
        at TaintExample.<clinit>(TaintExample.java)
java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 1
        at java.base/edu.columbia.cs.psl.phosphor.instrumenter.SourceTaintingMV.autoTaintArguments(SourceTaintingMV.java:87)
        at java.base/edu.columbia.cs.psl.phosphor.instrumenter.SourceTaintingMV.visitLdcInsn(SourceTaintingMV.java:66)
        at java.base/edu.columbia.cs.psl.phosphor.org.objectweb.asm.ClassReader.readCode(ClassReader.java:2398)
        at java.base/edu.columbia.cs.psl.phosphor.org.objectweb.asm.ClassReader.readMethod(ClassReader.java:1491)
        at java.base/edu.columbia.cs.psl.phosphor.org.objectweb.asm.ClassReader.accept(ClassReader.java:721)
        at java.base/edu.columbia.cs.psl.phosphor.org.objectweb.asm.ClassReader.accept(ClassReader.java:401)
        at java.base/edu.columbia.cs.psl.phosphor.SourceSinkTransformer.transform(SourceSinkTransformer.java:64)
        at java.base/edu.columbia.cs.psl.phosphor.PhosphorBaseTransformer.signalAndTransform(PhosphorBaseTransformer.java:21)
        at edu.columbia.cs.psl.jigsaw.phosphor.javaagent.PreMain$PhosphorTransformerBridge.transform(PreMain.java:36)
        at edu.columbia.cs.psl.jigsaw.phosphor.javaagent.PreMain$PhosphorTransformerBridge.transform(PreMain.java:26)
        at java.instrument/java.lang.instrument.ClassFileTransformer.transform(ClassFileTransformer.java:244)
        at java.instrument/sun.instrument.TransformerManager.transform(TransformerManager.java:188)
        at java.instrument/sun.instrument.InstrumentationImpl.transform(InstrumentationImpl.java:565)
        at java.instrument/sun.instrument.InstrumentationImpl.transform(InstrumentationImpl.java)
        at java.instrument/sun.instrument.InstrumentationImpl.retransformClasses0(Native Method)
        at java.instrument/sun.instrument.InstrumentationImpl.retransformClasses0(InstrumentationImpl.java)
        at java.instrument/sun.instrument.InstrumentationImpl.retransformClasses(InstrumentationImpl.java:168)
        at java.instrument/sun.instrument.InstrumentationImpl.retransformClasses(InstrumentationImpl.java)
        at edu.columbia.cs.psl.jigsaw.phosphor.javaagent.PreMain$1.retransformClasses(PreMain.java:68)
        at java.base/edu.columbia.cs.psl.phosphor.SourceSinkTransformer.retransform(SourceSinkTransformer.java:114)
        at TaintExample.<clinit>(TaintExample.java)
The sum of 12 and 12 is: 24
Sink
```

# Environment Details + Setup
## Java Version
```
$:~ java -version
openjdk version "16.0.1" 2021-04-20
OpenJDK Runtime Environment (build 16.0.1+9-Ubuntu-120.04)
OpenJDK 64-Bit Server VM (build 16.0.1+9-Ubuntu-120.04, mixed mode, sharing)
```

## JRE Instrumentation
```
$:~ java -jar phosphor-instrument-jigsaw/target/phosphor-instrument-jigsaw-0.1.0-SNAPSHOT.jar /usr/lib/jvm/java-16-openjdk-amd64/ jre-inst

/usr/lib/jvm/java-16-openjdk-amd64/bin/jlink -J-javaagent:/home/ahmad/Documents/VT/taintanalysis/phosphor/phosphor-instrument-jigsaw/target/phosphor-instrument-jigsaw-0.1.0-SNAPSHOT.jar -J--module-path=/home/ahmad/Documents/VT/taintanalysis/phosphor/phosphor-instrument-jigsaw/target/phosphor-instrument-jigsaw-0.1.0-SNAPSHOT.jar -J--add-modules=edu.columbia.cs.psl.jigsaw.phosphor.instrumenter -J--class-path= --output=jre-inst --phosphor-transformer=transform --add-modules=java.base,jdk.jdwp.agent,java.instrument,jdk.unsupported 
Embedding Phosphor from: /home/ahmad/Documents/VT/taintanalysis/phosphor/phosphor-instrument-jigsaw/target/phosphor-instrument-jigsaw-0.1.0-SNAPSHOT.jar
Processed: 1000
Processed: 2000
Processed: 3000
Processed: 4000
Processed: 5000
Processed: 6000
```

Kindly let me know if any more information is needed!

**Comment by jon-bell** [comment]
Hi,

I've reproduced this issue, created a failing integration test for it, and now fixed it in 21b0b87. If you run into any more issues, I am happy to help - the information that you provided was plenty to allow me to reproduce it, but in the future, it would be even easier if you might be able to add a failing integration test, like I added in 110e6f5 . Our test suite could definitely use significant improvement - but I am also happy to help construct the test cases as in this example, and appreciate the help identifying ways to improve it.

Somewhat related: I'm not sure what your goal/overall project is, but if you are interested in dynamic taint tracking for the JVM, [MirrorTaint](https://github.com/MirrorTaint/MirrorTaint) is described in [a recent ICSE 2023 article](https://lingming.cs.illinois.edu/publications/icse2023d.pdf) as incurring lower runtime overhead and having greater compatibility than Phosphor - you might want to try this out for your use-case, too.

Best,

Jon
